sql/handler.cc

```diff
@@ -1436,14 +1436,25 @@ int ha_prepare(THD *thd)
       }
     }
 
+unlog_xa_prepare:
     DEBUG_SYNC(thd, "at_unlog_xa_prepare");
-
     if (tc_log->unlog_xa_prepare(thd, all))
     {
       ha_rollback_trans(thd, all);
       error=1;
     }
   }
+  else if (thd->rgi_slave)
```

Comment:
Wouldn't it be better to OR this condition with l.1415 one?
That optimizes away the next `goto` and l.1439.